components/logins/src/db.rs

```diff
@@ -1235,4 +1268,108 @@ mod tests {
+            .get_by_base_domain(query)
+            .unwrap()
+            .into_iter()
+            .map(|l| l.hostname)
+            .collect::<Vec<String>>();
+        results.sort();
+        let mut sorted = expected.clone();
+        sorted.sort();
+        assert_eq!(sorted, results);
+    }
+
+    fn check_good_bad(
+        good: Vec<&str>,
+        bad: Vec<&str>,
+        good_queries: Vec<&str>,
+        zero_queries: Vec<&str>,
+    ) {
+        let db = LoginDb::open_in_memory(Some("testing")).unwrap();
+        for h in good.iter().chain(bad.iter()) {
+            db.add(Login {
+                hostname: (*h).into(),
+                http_realm: Some((*h).into()),
+                password: "test".into(),
+                ..Login::default()
+            })
+            .unwrap();
+        }
+        for query in good_queries {
+            check_matches(&db, query, &good);
+        }
+        for query in zero_queries {
+            check_matches(&db, query, &vec![]);
+        }
+    }
+
+    #[test]
+    fn test_get_by_base_domain_invalid() {
+        check_good_bad(
+            vec!["https://example.com"],
+            vec![],
+            vec![],
+            vec!["invalid query"],
+        );
+    }
+
+    #[test]
+    fn test_get_by_base_domain() {
+        check_good_bad(
+            vec![
+                "https://example.com",
+                "https://www.example.com",
+                "http://www.example.com",
+                "http://www.example.com:8080",
+                "http://sub.example.com:8080",
+                "https://sub.example.com:8080",
+                "https://sub.sub.example.com",
+            ],
+            vec![
+                "https://badexample.com",
+                "https://example.co",
+                "https://example.com.au",
+            ],
+            vec!["example.com"],
+            vec!["foo.com"],
+        );
+        // punycode! This is likely to need adjusting once we normalize
+        // on insert.
+        check_good_bad(
+            vec![
+                "http://😍.com",
+                "http://xn--r28h.com", // punycoded version of the above.
+            ],
+            vec!["http://💖.com"],
+            vec!["😍.com", "xn--r28h.com"],
+            vec![],
+        );
+    }
+
+    #[test]
+    fn test_get_by_base_domain_ipv4() {
+        check_good_bad(
+            vec!["http://127.0.0.1", "https://127.0.0.1:8000"],
+            vec!["https://127.0.0.0", "https://example.com"],
+            vec!["127.0.0.1"],
+            vec!["127.0.0.2"],
+        );
+    }
+
+    #[test]
+    fn test_get_by_base_domain_ipv6() {
+        check_good_bad(
+            vec![
+                "http://[::1]",
+                "https://[::1]:8000",
+                "https://[0:0:0:0:0:0:0:1]", // this is [::1]
+            ],
+            vec!["https://[0:0:0:0:0:0:1:1]", "https://example.com"],
+            vec!["[::1]", "[0:0:0:0:0:0:0:1]"],
+            vec!["[0:0:0:0:0:0:1:2]"],
```

Comment:
I had thought that an IPv6 hostnames shouldn't include the square brackets, at least `nsIURI.asciiHost`, `nsIURI.displayHost` don't return them but then I found that `URL.hostname` does… (Note that `nsIURI.*Host` <=> `URL.hostname`, whereas `nsIURI.*HostPort` <=> `URL.host`). If you found that `Host::parse` can't handle IPv6 addresses without the square brackets then I can handle this on the toolkit side but this should be documented in the API docs too. I does seem like the Rust URL library is conflating "host" and "hostname" though… Unfortunately they don't have `Hostname::parse` which is what you actually want.